helix-core/src/comment.rs

```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
     Transaction::change(doc, changes.into_iter())
 }
 
+fn find_last_non_whitespace_char(text: RopeSlice) -> Option<usize> {
+    (0..text.len_chars()).rev().find(|&i| {
+        text.get_char(i)
+            .map(|ch| !ch.is_whitespace())
+            .unwrap_or(false)
+    })
+}
+
+#[derive(Debug, PartialEq)]
+struct CommentChange {
+    range: Range,
+    open_pos: usize,
+    close_pos: usize,
+    open_margin: usize,
+    close_margin: usize,
+}
+
+fn find_block_comments(
+    open: &str,
+    close: &str,
+    text: RopeSlice,
+    selection: &Selection,
+) -> (bool, Vec<CommentChange>) {
```

Comment:
this could return an iterator I think which would be more efficient/idiomatic

Reply:
I hope I addressed everything else but sadly I don't think we can use an iterator here because we need to check all the selections before we know if we are commenting or not. See how we clear the clear the vec if we find an uncommented selection so that we can normalize all the selections to all be commented.

Reply:
Ah yeah ofcourse I didn't think j about that my bad